burstphoto/texture/texture.swift

```diff
@@ -157,14 +164,95 @@ func blur(_ in_texture: MTLTexture, with_pattern_width mosaic_pattern_width: Int
 }
 
 
+/// Calculate the sum of each subpixel in the mosaic pattern within the specified region.
+/// Returns and array of length mosaic\_pattern\_width^2 which is to be treated as a flattened 2D array: 2D(x, y) == 1D(x + width\*y)
+func calculate_black_levels(for texture: MTLTexture, from_masked_areas masked_areas: UnsafeMutablePointer<Int32>, mosaic_pattern_width: Int) -> [Int] {
+    var num_pixels: Float = 0.0
+    var command_buffers: [MTLCommandBuffer] = []
+    var black_level_buffers: [MTLBuffer] = []
+    var black_level_from_masked_area = [Float](repeating: 0.0, count: Int(mosaic_pattern_width*mosaic_pattern_width))
+    
+    for i in 0..<4 {
+        // Up to 4 masked areas exist, as soon as we reach -1 we know there are no more masked areas after.
+        if masked_areas[4*i] == -1 { break }
+        let top     = masked_areas[4*i + 0]
+        let left    = masked_areas[4*i + 1]
+        let bottom  = masked_areas[4*i + 2]
+        let right   = masked_areas[4*i + 3]
+        
+        num_pixels += Float((bottom - top) * (right - left))
+        
+        // Create output texture from the y-axis blurring
+        let texture_descriptor = MTLTextureDescriptor()
+        texture_descriptor.textureType = .type2D
+        texture_descriptor.pixelFormat = .r32Float
+        texture_descriptor.width = Int(right - left)
+        texture_descriptor.height = mosaic_pattern_width
+        texture_descriptor.usage = [.shaderRead, .shaderWrite]
+        let summed_y = device.makeTexture(descriptor: texture_descriptor)!
+        
+        // Sum along columns
+        let command_buffer = command_queue.makeCommandBuffer()!
+        let command_encoder = command_buffer.makeComputeCommandEncoder()!
+        command_buffer.label = "Black Levels \(i) for \(texture.label)"
+        command_encoder.setComputePipelineState(sum_rect_columns_state)
+        let thread_groups_per_grid = MTLSize(width: summed_y.width, height: summed_y.height, depth: 1)
+        let max_threads_per_thread_group = sum_rect_columns_state.maxTotalThreadsPerThreadgroup
+        let threads_per_thread_group = MTLSize(width: max_threads_per_thread_group, height: 1, depth: 1)
```

Comment:
@chris-rank, I'm not clear on how to properly pick values for these.

Reply:
I think you manually set the values equivalent to the function "get_threads_per_thread_group" provided in Burst Photo. As far as I know, this function is used everywhere throughout the app. It was already included when I started contributing to the app and I never questioned if it is the optimal distribution of threads.